src/host/posix/dnssd.hpp

```diff
@@ -112,25 +129,243 @@ class DnssdPlatform : public Mdns::StateObserver, private NonCopyable
+            }
+        }
+
+        bool IsEmpty(void) { return mInfraIfCallbackTable.empty(); }
+
+    private:
+        typedef std::map<DiscoveryCallbackIdentifier, CallbackType> CallbackMap;
+
+        std::map<uint64_t, CallbackMap> mInfraIfCallbackTable;
+    };
+
     void HandleMdnsState(Mdns::Publisher::State aState) override;
 
     void                            UpdateState(void);
     Mdns::Publisher::ResultCallback MakePublisherCallback(RequestId aRequestId, RegisterCallback aCallback);
 
     static std::string KeyNameFor(const Key &aKey);
 
+    static void HandleDiscoveredService(const std::string &aType, const Mdns::Publisher::DiscoveredInstanceInfo &aInfo);
+    static void HandleDiscoveredHost(const std::string &aHostName, const Mdns::Publisher::DiscoveredHostInfo &aInfo);
+
+    void ProcessServiceBrowsers(const std::string &aType, const Mdns::Publisher::DiscoveredInstanceInfo &aInfo);
+    void ProcessServiceResolvers(const std::string &aType, const Mdns::Publisher::DiscoveredInstanceInfo &aInfo);
+    void ProcessTxtResolvers(const std::string &aType, const Mdns::Publisher::DiscoveredInstanceInfo &aInfo);
+    void ProcessAddrResolvers(const std::string                                        &aHostName,
+                              const Mdns::Publisher::DiscoveredHostInfo                &aInfo,
+                              std::map<DnsName, InfraIfCallbackTable<AddressCallback>> &aResolversTable);
+    void ProcessIp6AddrResolvers(const std::string &aHostName, const Mdns::Publisher::DiscoveredHostInfo &aInfo);
+    void ProcessIp4AddrResolvers(const std::string &aHostName, const Mdns::Publisher::DiscoveredHostInfo &aInfo);
+
+    void StartAddressResolver(const AddressResolver                                    &aAddressResolver,
+                              const AddressCallback                                    &aCallback,
+                              const DiscoveryCallbackIdentifier                        &aId,
+                              std::map<DnsName, InfraIfCallbackTable<AddressCallback>> &aResolversTable);
+    void StopAddressResolver(const AddressResolver                                    &aAddressResolver,
+                             const DiscoveryCallbackIdentifier                        &aId,
+                             std::map<DnsName, InfraIfCallbackTable<AddressCallback>> &aResolversTable);
+
     static DnssdPlatform *sDnssdPlatform;
 
-    Mdns::Publisher         &mPublisher;
-    State                    mState;
-    bool                     mRunning;
-    Mdns::Publisher::State   mPublisherState;
-    DnssdStateChangeCallback mStateChangeCallback;
+    Mdns::Publisher                                               &mPublisher;
+    State                                                          mState;
+    bool                                                           mRunning;
+    bool                                                           mInvokingCallbacks;
```

Comment:
I saw this used for blocking the future discoveries. Could you explain why it's needed?

Reply:
I'm not so sure whether there is a use case. Could there be a browse of services triggering following up resolutions of records?

Reply:
It's not a problem if we really have such use cases. But we just need to be clear on the API's capability.

Since the caller of these dnssd platform APIs are OT core dnssd, the callback is also implemented in OT core. We can post a task to do trailing resolution. Thus the `Start` won't be directly invoked in the callback.